Convert hui-icon-element to TypeScript/LitElement

iantrich opened the pull request:
extract ElementClick mixin functions

- `computeTooltip` only works when called directly in `hui-icon-element` but doesn't when I try to replace what is in the existing `ElementClick` mixin.
- waiting on new navigate function to finish extracting handleClick function and creation of service data needs to be cleaned up for TS linter to like it.

iantrich committed:
Convert hui-icon-element to TypeScript/LitElement and extract ElementClick mixin functions

iantrich committed:
Pass correct node to handleClick

zsarnett commented on src/common/dom/handle-click.ts:
```diff
@@ -0,0 +1,44 @@
+    case "call-service": {
+      if (config.service) {
+        const [domain, service] = config.service.split(".", 2);
+        const serviceData = Object.assign(
```
Checkout this notation `{entity_id: config.entity, ...config.service_data}`

zsarnett commented on src/common/dom/handle-click.ts:
```diff
@@ -0,0 +1,44 @@
+    return;
+  }
+
+  switch (action) {
```
Should there be a default case?

iantrich replied:
more-info?

zsarnett replied:
Bingo

iantrich replied:
This is actually already handled when action is set:

`let action = config.tap_action || "more-info";`

iantrich committed:
Remove element-click refactor

Was unable to get computeTooltip to work for non-converted elements. There are not many elements that use this so will just remove this mixin once conversion of all to TS has been completed.

iantrich committed:
Implemented navigate function

Fix linting error

iantrich commented:
This should be good to go now, I believe

iantrich committed:
cleaned up elementconfig type order

iantrich commented:
Scratch that. Long press is not working

iantrich committed:
Add return types

iantrich commented:
Scratch the scrath :) I didn't realize that long press was for mobile online. The timing is quite finicky to test.

balloob commented on src/common/dom/handle-click.ts:
```diff
@@ -0,0 +1,44 @@
+export const handleClick = (
+  node: HTMLElement,
+  hass: HomeAssistant,
+  config: LovelaceElementConfig,
```
This belongs in lovelace element folder. Nothing common about it.

balloob commented on src/common/string/compute-tooltip.ts:
```diff
@@ -0,0 +1,38 @@
+
+export const computeTooltip = (
+  hass: HomeAssistant,
+  config: LovelaceElementConfig
```
Same.
